src/proposed.api.d.ts

```diff
@@ -0,0 +1,99 @@
+    token: string;
+    /**
+     * Authorization header to be used when connecting to the server.
+     */
+    authorizationHeader?: Record<string, string>;
+    /**
+     * The local directory that maps to the remote directory of the Jupyter Server.
+     * E.g. assume you start Jupyter Notebook with --notebook-dir=/foo/bar,
+     * and you have a file named /foo/bar/sample.ipynb, /foo/bar/sample2.ipynb and the like.
+     * Then assume the mapped local directory will be /users/xyz/remoteServer and the files sample.ipynb and sample2.ipynb
+     * are in the above local directory.
+     *
+     * Using this setting one can map the local directory to the remote directory.
+     * In this case the value of this property would be /users/xyz/remoteServer.
+     *
+     * Note: A side effect of providing this value is the session names are generated the way they are in Jupyter Notebook/Lab.
+     * I.e. the session names map to the relative path of the notebook file.
+     * As a result when attempting to create a new session for a notebook/file, Jupyter will
+     * first check if a session already exists for the same file and same kernel, and if so, will re-use that session.
+     */
+    mappedRemoteNotebookDir?: string;
+    /**
+     * Returns the sub-protocols to be used. See details of `protocols` here https://developer.mozilla.org/en-US/docs/Web/API/WebSocket/WebSocket
+     * Useful if there is a custom authentication scheme that needs to be used for WebSocket connections.
+     * Note: The client side npm package @jupyterlab/services uses WebSockets to connect to remote Kernels.
+     */
+    webSocketProtocols?: string[];
+}
+
+export type ProviderResult<T> = T | undefined | null | Promise<T | undefined | null>;
+
+export interface JupyterServerProvider {
+    /**
+     * Needs to be unique and the same across sessions.
+     */
+    readonly id: string;
+    /**
+     * Display name.
+     */
+    readonly name: string;
+    /**
+     * Detailed description.
+     */
+    readonly detail: string;
+    /**
+     * List of all known servers.
+     */
+    readonly servers: JupyterServer[];
+    /**
+     * Triggered when servers are added/removed.
+     */
+    onDidChangeServers?: Event<void>;
+    /**
+     * Called when the auth information needs to be populated for a server.
+     */
+    resolveServer(server: JupyterServer): Promise<JupyterServer>;
+    /**
+     * Called when a server is to be removed from the list of servers.
+     */
+    removeServer(server: JupyterServer): Promise<void>;
+    /**
+     * Provides full control of the quick pick to the extension, so that it can add additional buttons, etc.
+     * The result of this promise should be the selected server or undefined if no server was selected.
+     *
+     * If nothing is selected, then the quick pick reverts back to the previous View before this method was invoked.
+     */
+    handleQuickPick?(quickPick: QuickPick<QuickPickItem>): ProviderResult<JupyterServer>;
```

Comment:
Giving full access to the QuickPick is the easiest
* Today we have custom arguments for `back`, 
* custom handling for when something is selected we get a return value of the handle and then from that we get the server info
* We have custom handling for selecting the default quick pick item
* We have requirements to change this for user uri picker (https://github.com/microsoft/vscode-jupyter/issues/13892)

All of the above is very easy if we just provide the quick pick as is,
All we need to do is, once the promise resolves with `undefined`, then we need to restore the quick pick state back to the previous view
& thats easy
* We just update the items, selections, title, buttons, etc (basically the same as running some existing code that builds the quick pick)

Reply:
For reference this is the existing API
* We can remove concept of `handles`, and have a simpler API

https://github.com/microsoft/vscode-jupyter/blob/285988fa81fd56598bffea06dfa86a5e2ab97a2f/src/api.d.ts#L74-L110

Reply:
We discussed offline about one solution to simplify the creation of `JupyterServer` by making `JupyterServer` a managed object by Jupyter extension.

* 3rd party extension calls `JupyterAPI.createServerProvider(id, displayName)` to get a managed `ServerProvider` object
* call `serverProvider.createServer(id, displayName, url, authHandler)` to create a `JupyterServer` for this provider. Once created, the server will be visible to users on the kernel picker source
* `JupyerServer` is an object managed by Jupyter extension. It can expose `JupyterServer.serverConnection` property, which is a JupyterLab api object.
* The `JupyterServer` can be removed by either the Jupyter extension or the 3rd party extension.

3rd party extension will use `ServerProvider.createServer` to indicate there are servers available, they no longer need to use `EventEmitters` (`onDidChangeServers`). The `JupyterAPI.addRemoteJupyterServer` can also be removed in favor of this.